core/src/Streamly/Internal/Data/Stream/TH.hs

```diff
@@ -0,0 +1,542 @@
+                        []
+                  ]
+            ]
+    reader <-
+        instanceD
+            (sequence
+                 [ appT (conT _Monad) (appT (conT _Type) (varT _m))
+                 , appT (appT (conT _MonadReader) (varT _r)) (varT _m)
+                 ])
+            (appT
+                 (appT (conT _MonadReader) (varT _r))
+                 (appT (conT _Type) (varT _m)))
+            [ pragInlD _ask Inline FunLike AllPhases
+            , funD _ask [clause [] (normalB (appE (varE _lift) (varE _ask))) []]
+            , pragInlD _local Inline FunLike AllPhases
+            , funD
+                  _local
+                  [ clause
+                        [varP _f, conP _Type [varP _strm]]
+                        (normalB
+                             (appE
+                                  (conE _Type)
+                                  (appE
+                                       (appE (varE _local) (varE _f))
+                                       (varE _strm))))
+                        []
+                  ]
+            ]
+    return [trans, io, throw, reader]
+
+    where
+
+    _Type = mkName dtNameStr
+    _unType = mkName (unTypeStr dtNameStr)
+
+--------------------------------------------------------------------------------
+-- Functions
+--------------------------------------------------------------------------------
+
+-- | Create a type with a zip-like applicative.
+--
+-- __Create a type with a serial zipping applicative property:__
+--
+-- @
+-- import Control.Monad.Catch (MonadThrow(..))
+-- import Control.Monad.IO.Class (MonadIO(..))
+-- import Control.Monad.Reader.Class (MonadReader(..))
+-- import Control.Monad.Trans.Class (MonadTrans(..))
+-- import Streamly.Data.Stream (Stream)
+--
+-- import qualified Streamly.Data.Stream
+-- import qualified Streamly.Data.Stream as Stream
+-- import qualified Streamly.Internal.Data.Stream.TH as StreamTH
+--
+-- applySerial :: Monad m => Stream m (a -> b) -> Stream m a -> Stream m b
+-- applySerial = Stream.zipWith ($)
+--
+-- \$(StreamTH.makeStreamZipApplicative \"ZipSerial\" "applySerial" False)
+-- @
+--
+-- __Create a type with a concurrent zipping applicative property:__
+--
+-- @
+-- import Control.Monad.Catch (MonadThrow(..))
+-- import Control.Monad.IO.Class (MonadIO(..))
+-- import Control.Monad.Reader.Class (MonadReader(..))
+-- import Control.Monad.Trans.Class (MonadTrans(..))
+-- import Streamly.Data.Stream (Stream)
+--
+-- import qualified Streamly.Data.Stream
+-- import qualified Streamly.Data.Stream.Concurrent
+-- import qualified Streamly.Data.Stream.Concurrent as Concur
+-- import qualified Streamly.Internal.Data.Stream.TH as StreamTH
+--
+-- applyConcurrent :: Monad m => Stream m (a -> b) -> Stream m a -> Stream m b
+-- applyConcurrent = Concur.zipWith ($)
+--
+-- \$(StreamTH.makeStreamZipApplicative \"ZipConcurrent\" "applyConcurrent" True)
+-- @
+--
+makeStreamZipApplicative
```

Comment:
Can we make it simpler  like `mkZipType`? Stream is implicit.

Reply:
Type is also implicit?

Reply:
`mkZipAp` and `mkMonad`?